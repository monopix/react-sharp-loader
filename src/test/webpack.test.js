const vm = require('vm')
const fs = require('fs')
const _webpack = require('webpack')
const path = require('path')
const MemoryFileSystem = require('memory-fs')
const image = fs.readFileSync(`${process.env.PWD}/src/test/image.jpg`)
const imageResult = fs.readFileSync(`${process.env.PWD}/src/test/result.jpg`)

const { PWD } = process.env
const resolvePath = (...args) => path.resolve(PWD, path.join(...args))

jest.setTimeout(15000)

const webpackConfig = (query, entry = 'index.js', extra) => {
  return {
    mode: 'production',
    // entry: path.join(__dirname, '..', '..', 'example', entry),
    entry: resolvePath('example', entry),
    context: resolvePath('example'),
    output: {
      path: resolvePath('dist'),
      publicPath: '/foo',
      filename: 'bundle.js',
      libraryTarget: 'commonjs2',
    },
    module: {
      rules: [
        {
          test: /\.(gif|jpe?g|png|svg|tiff)(\?.*)?$/,
          use: {
            loader: resolvePath('src', 'index.js'),
            query,
          },
        },
      ],
    },
    ...extra,
  }
}

const webpack = (options, inst, extra) => {
  const configuration = webpackConfig(options, inst, extra)
  const compiler = _webpack(configuration)
  compiler.outputFileSystem = new MemoryFileSystem()

  return new Promise(resolve => {
    compiler.run((err, _stats) => {
      expect(err).toBe(null)

      if (_stats.hasErrors()) {
        // eslint-disable-next-line
        console.log(_stats.toString())
        // throw new Error('webpack error occured')
      }

      const stats = _stats.toJson()
      const files = {}
      let code = ''

      stats.assets.forEach(asset => {
        files[asset.name] = compiler.outputFileSystem.readFileSync(
          path.join(configuration.output.path, asset.name)
        )
        if (asset.name === 'bundle.js') {
          code = files[asset.name].toString('utf8')
        }
      })

      const sandbox = vm.createContext({})
      sandbox.global = {}
      sandbox.module = { exports: {} }
      vm.runInContext(code, sandbox)

      resolve({ stats, files, exports: sandbox.module.exports })
    })
  })
}

xit('should works with a webpack and emit proper files', () => {
  const options = {
    cache: false,
    config: {
      grayscale: true,
    },
    // presets: {
    //   default: [{ width: 100 }],
    //   thumbnail: [{ width: [150, 300] }],
    //   prefetch: [{ width: [50, 200], grayscale: [false] }],
    //   fallback: [{ width: 240 }],
    //   placeholder: [{ width: 30, base64: true }],
    // },
    // outputs: 'default',
    // fallback: 'fallback',
    // placeholder: 'placeholder',
  }
  return webpack(options).then(({ stats, files }) => {
    expect(stats).not.toBe(null)
    expect(stats.assets.length).toBe(2)
    // NOTE: The order isn't guaranteed
    expect(Buffer.compare(files[stats.assets[0].name], imageResult)).toBe(1)
  })
})

xit('should works with a webpack and exports react module', () => {
  const options = {
    cache: false,
    config: {
      grayscale: true,
    },
    presets: {
      prefetch: {
        srcset: [
          {
            format: ['webp', 'jpg'], // order have a meaning
            width: [60, 150],
          },
        ],
        fallback: [{ width: 80 }],
        placeholder: [{ width: 50, base64: true }],
      },
    },

  }
  return webpack(options).then(({ exports }) => {
    expect(exports.a()).toMatchSnapshot()
  })
})
