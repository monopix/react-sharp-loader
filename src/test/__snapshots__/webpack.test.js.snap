// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should works with a webpack and exports react module 1`] = `
<div
  style={
    Object {
      "overflow": "hidden",
      "paddingBottom": "78%",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <img
    src="data:image/jpeg;base64,/9j/2wBDAAYEBQYFBAYGBQYHBwYIChAKCgkJChQODwwQFxQYGBcUFhYaHSUfGhsjHBYWICwgIyYnKSopGR8tMC0oMCUoKSj/2wBDAQcHBwoIChMKChMoGhYaKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCj/wAARCAAnADIDASIAAhEBAxEB/8QAGgAAAwEBAQEAAAAAAAAAAAAAAAQGBwUBAv/EAC4QAAEDAwMCBQQBBQAAAAAAAAECAwQABREGEiEiQQcTFDFhFSNRcTIWU5Khsf/EABQBAQAAAAAAAAAAAAAAAAAAAAD/xAAUEQEAAAAAAAAAAAAAAAAAAAAA/9oADAMBAAIRAxEAPwDheGumId1dtapcdKmfTSHXiEBRWsOANjB5PBPtzgZ7Ut4raXjwGLVItzLSVKZWZPlpSOvdxnbxwMCn9PH09t+lzUx8oSpJPrS0pO8ggjoODxwc85pbVu+2WCHEQlIipLjaFB7zCVZ3HPSnHv8Aig63hGbX9JjR5tkXc5K3Fq+2nKkoCjk4yM/9+atri/o9iGZarVbyFqCGkqLqFKPPYE49qy/RtxetElgekwtorILkrykkKVnkbTVZJdiSm0rVFtyHE8gpu2Dn/CgzvxEEaXckTbbGTGiOspWhtOcAFSh357d6ptO6ful0v2onoUe3pt1skKbPnxmyMAnpThOSQBnHf80vfIQnRFhtCMMsoaQG3w7naVEncAB3rRbGuI1ZNVQ0i4MO3VxxxDjcNY8lSsjKsq98ccfNB2LdpbTci3xn3LTAUtxpK1KEdIySAc+1FfMO8xY8RhkMTVhtCUBXp1DOBjOKKDBnL6yzdSLgwt2KpKNy0jray0MFBBHO7Bwfx+6595nuz9KRVPLWtQmPdSsZI2JAzj4ApZT9nkISqV9UDxCErDRbCCBxwD/rNezpVrENiLbm5i2kuLcX6spB6hjgoPxQWVmM2439mJHkpQ7u81lITkrwAduTx7AnFUXibNh2FEe3xlRiZTaxIKW9ysHbv6hjaQrIGM8Z55rObfq68RCBEltMgAJBEdonH4JKcmlL/eJ16ejLuMhL6mklCCltKNoJzjpA70F1e5UJzTFqTavttFCd6QonCwVAhR7nGOT78Uo7qly3365x0SXDGdmrDragBlOV8Djuojn4+amosnZZmY6FrUULLhSoJ/kcexAz27mm7zKsE156Qq13Zl90lSimW1tySSTjZ80GwQbsFQo6nFYWW0lX7xRWTtaxbZaQ03Ee2ISEp3OpJwOOeKKBeHdIXG62Rin8kVSwdYWeHBXFd0zantwJLjjRKjn5FFFAl/UUJTyExrHblew6kkGuZrSbDnzISY8NmIttDpWW0/yATkDH770UUFNom62mHpphCojbj5QXXFOo3ZUfcD4rou6jtjkUuNRIyVAcnyM80UUE+u/tlaj6OJyf7NFFFB//2Q=="
    style={
      Object {
        "height": "100%",
        "position": "absolute",
        "right": 0,
        "top": 0,
        "verticalAlign": "middle",
        "width": "100%",
      }
    }
  />
  <picture>
    <source
      sizes="100vw"
      srcSet="__webpack_public_path__/test.7f4660ea.webp 60w, __webpack_public_path__/test.20510907.webp 150w"
      type="image/webp"
    />
    <img
      sizes="100vw"
      src="__webpack_public_path__/test.346cec82.jpeg"
      srcSet="__webpack_public_path__/test.92b92d3c.jpeg 60w, __webpack_public_path__/test.3a54733d.jpeg 150w"
      style={
        Object {
          "height": "100%",
          "position": "absolute",
          "right": 0,
          "top": 0,
          "verticalAlign": "middle",
          "width": "100%",
        }
      }
    />
  </picture>
</div>
`;
